api/server.js

```diff
@@ -41,41 +44,42 @@ process
     logger.info('Extracting initial documents…');
     await uploadDefaultDocs.run();
     logger.info('Extracting initial documents completed successfully');
-    
+
     logger.info('Loading configuration…');
     await config.load();
     logger.info('Configuration loaded successfully');
     await config.listen();
-    
+
     logger.info('Merging translations…');
     await translations.run();
     logger.info('Translations merged successfully');
-    
+
     logger.info('Running db migrations…');
     await migrations.run();
     logger.info('Database migrations completed successfully');
 
     logger.info('Updating xforms…');
     await generateXform.updateAll();
     logger.info('xforms updated successfully');
+
+    // Define error-handling middleware last.
+    // http://expressjs.com/guide/error-handling.html
+    app.use((err, req, res, next) => {
+      if (res.headersSent) {
+        // If we've already started a response (eg streaming), pass on to express to abort it
+        // rather than attempt to resend headers for a 5xx response
+        return next(err);
+      }
+      serverUtils.serverError(err, req, res);
+    });
+
+    app.listen(apiPort, () => {
+      logger.info('Medic API listening on port ' + apiPort);
+    });
```

Comment:
I don't think this should be in the `try/catch` block. Generally I try to put as little in there as possible in order to make the `catch` specific to one kind of error.

I don't think this block should throw anything anyway so it can be reverted to where it was before.

Reply:
Yes! I wanted to keep as little as possible in the `try/catch` block too.  However, I ended up with it all in there because of a scoping issue that I didn't understand.  When I move these three lines to be just after the `catch`, I get this error:

```
2020-09-01 09:35:46 ERROR: UNHANDLED REJECTION! 
2020-09-01 09:35:46 ERROR:   Reason: ReferenceError: app is not defined
    at /home/mrjones/Documents/MedicMobile/cht-core/api/server.js:82:3 {
  [stack]: 'ReferenceError: app is not defined\n' +
    '    at /home/mrjones/Documents/MedicMobile/cht-core/api/server.js:82:3',
  [message]: 'app is not defined'
} 
```

which is super weird to me and wanted to ask: do `const` declarations inside the the `try{}` block have a limited scope such that calls outside it can't access those constants?  This seems weird to me!

And, now that this PR is out of draft, on this same topic, I wanted to ask a larger question: is it safe to move all these `require()` and `await` calls inside the the `try{}` block? While I'm still debugging the last errors in travis, everything _seems_ to be working! The reason that I did this, aside from the scoping issue I listed above, is that it makes for a really nice and tidy error when you try and start api/sentinel and the couch server can't be reached:

```
2020-08-28 15:57:55 ERROR: Fatal error initialising medic-api
2020-08-28 15:57:55 ERROR: "CouchDB doesn't seem to be running on http://mrjones-is-great:5984/medic. Tried to connect but got an error:\n" +
  ' Error: getaddrinfo ENOTFOUND mrjones-is-great\n' +
  '    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:66:26)' 
```

This is SO much nicer than what it was before with [two uncaught exceptions](https://github.com/medic/cht-core/issues/6376#issuecomment-680359369).

Reply:
You're right about the scoping. Old school JS used `var` which was _function scoped_ which meant this would work just fine, but `const` and `let` are _block scoped_ which is much more like other programming languages and much less likely to cause unexpected bugs. Here's a blog post which discusses the difference: https://www.freecodecamp.org/news/var-let-and-const-whats-the-difference/

I'm not sure why the server checks code runs before the requires - we expect the requires to always succeed and be reasonably fast so I would move that out of the try/catch and put it above the serverChecks call. That way it'll be scoped correctly.

It's definitely nicer to have the catch around the serverChecks - uncaught errors or promise rejections are always a mistake!

Reply:
Ah - this was awesome - thanks! I'd read that freecodecamp.org post before, but clearly not close enough.  I pulled the `require()` and `log()` calls out of the `try/catch` after implementing helpful feedback in 27b391a.